glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,226 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+//! Module for registering shared types for Rust types.
+
+use crate::translate::*;
+
+pub trait RefCounted: Clone + Sized + 'static {
+    /// The inner type
+    type InnerType;
+
+    /// The function used to increment the inner type refcount
+    unsafe fn ref_(this: *const Self::InnerType) -> *const Self::InnerType;
+
+    /// Converts the RefCounted object to a raw pointer to InnerType
+    unsafe fn into_raw(self) -> *const Self::InnerType;
+
+    /// Converts a raw pointer to InnerType to a RefCounted object
+    unsafe fn from_raw(this: *const Self::InnerType) -> Self;
+}
+
+impl<T> RefCounted for std::sync::Arc<T>
+where
+    T: 'static,
+{
+    type InnerType = T;
+
+    unsafe fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+        std::sync::Arc::increment_strong_count(this);
+        this
+    }
+
+    unsafe fn into_raw(self) -> *const Self::InnerType {
+        std::sync::Arc::into_raw(self)
+    }
+
+    unsafe fn from_raw(this: *const Self::InnerType) -> Self {
+        std::sync::Arc::from_raw(this)
+    }
+}
+
+impl<T> RefCounted for std::rc::Rc<T>
+where
+    T: 'static,
+{
+    type InnerType = T;
+
+    unsafe fn ref_(this: *const Self::InnerType) -> *const Self::InnerType {
+        use std::mem::ManuallyDrop;
+        let this_rc = ManuallyDrop::new(std::rc::Rc::from_raw(this));
+        std::rc::Rc::into_raw(ManuallyDrop::take(&mut this_rc.clone()))
+    }
+
+    unsafe fn into_raw(self) -> *const Self::InnerType {
+        std::rc::Rc::into_raw(self)
+    }
+
+    unsafe fn from_raw(this: *const Self::InnerType) -> Self {
+        std::rc::Rc::from_raw(this)
+    }
+}
+
+/// Trait for defining shared types.
+///
+/// Links together the type name with the type itself.
+///
+/// See [`register_shared_type`] for registering an implementation of this trait
+/// with the type system.
+///
+/// [`register_shared_type`]: fn.register_shared_type.html
+pub trait SharedType: Clone + Sized + 'static {
+    /// Shared type name.
+    ///
+    /// This must be unique in the whole process.
+    const NAME: &'static str;
+
+    /// The inner refcounted type
+    type RefCountedType: RefCounted;
```

Comment:
```suggestion
    type RefCountedType: RefCounted<InnerType = Self>;
```

Is what you would do here in the current situation, or

```suggestion
    type RefCountedType: RefCounted<Self>;
```

if you make it a generic parameter. That would ensure that the types are actually matching up properly.